src/ucp/rndv/proto_rndv.c

```diff
@@ -631,14 +636,179 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+                "%s rts->address=0x%" PRIx64 " rkey_length=%zu", title, address,
+                rkey_length);
+}
+
+static ucs_status_t
+ucp_proto_rndv_unpack_contig_rkey(const void *buffer, size_t length,
+                                  ucp_ep_h ep, ucp_ep_config_t *ep_config,
+                                  ucp_request_t *req)
+{
+    ucp_rkey_h rkey;
+    ucs_status_t status;
+
+    /* Do not unpack rkeys from MDs with rkey_ptr capability, except
+     * rkey_ptr_lane. Examples are: sysv and posix. Such keys, if packed,
+     * are unpacked only once and cached in the peer_mem hash on ep. It is
+     * done by the specific protocols (if selected) which use them.
+     */
+    status = ucp_ep_rkey_unpack_internal(
+            ep, buffer, length, ep_config->key.reachable_md_map,
+            ep_config->rndv.proto_rndv_rkey_skip_mds, &rkey);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    req->send.rndv.rdata_count = 0;
+    req->send.rndv.rkey        = rkey;
+    /* Caching rkey_buffer pointer for later unpacking of shm keys in
+     * rkey_ptr mtype ppln protocol. */
+    req->send.rndv.rkey_buffer = buffer;
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE const void *
+ucp_proto_rndv_unpack_iov_count(const void *p, size_t *iov_count)
+{
+    *iov_count = *(const size_t*)p;
+    return UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
+}
+
+static UCS_F_ALWAYS_INLINE const void *
+ucp_proto_rndv_unpack_iov_rkey_buffer(const void *p, const void **rkey_buffer,
+                                      size_t *rkey_length,
+                                      uint64_t *remote_address,
+                                      size_t *remote_size)
+{
+    *remote_address = *(const uint64_t*)p;
+    p               = UCS_PTR_BYTE_OFFSET(p, sizeof(uint64_t));
+
+    *remote_size = *(const size_t*)p;
+    p            = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
+
+    *rkey_length = *(const size_t*)p;
+    p            = UCS_PTR_BYTE_OFFSET(p, sizeof(size_t));
+
+    *rkey_buffer = p;
+    p            = UCS_PTR_BYTE_OFFSET(p, *rkey_length);
+
+    return p;
+}
+
+static ucs_status_t
+ucp_proto_rndv_unpack_iov_rkeys(const void *buffer, size_t length, ucp_ep_h ep,
+                                ucp_ep_config_t *ep_config, ucp_request_t *req)
+{
+    const void *p = buffer;
+#if UCS_ENABLE_ASSERT
+    const void *p_end = UCS_PTR_BYTE_OFFSET(p, length);
+#endif
+    size_t iov_count, iov_index;
+    ucp_rkey_h rkey;
+    const void *rkey_buffer;
+    size_t rkey_length;
+    uint64_t raddress;
+    size_t rsize, accumulate_size;
+    ucs_status_t status;
+
+    p = ucp_proto_rndv_unpack_iov_count(p, &iov_count);
+    if (iov_count == 0) {
+        return UCS_ERR_NO_MESSAGE;
+    }
+
+    req->send.rndv.rdata = ucs_malloc(iov_count * sizeof(*req->send.rndv.rdata),
+                                      "proto_rndv");
+    if (req->send.rndv.rdata == NULL) {
+        ucs_error("failed to allocate rma array");
```

Comment:
`array for iov rkeys`?

Reply:
Yes, but not only rkeys, but also iov length and address.